### PR TITLE
Special case `str` dtype in array creation

### PR DESCRIPTION
**Pull request by TomAugspurger**
Special cases `str`, interpreting it as `np.dtype("object")`, to match zarr-python 2.x's behavior.

Closes https://github.com/zarr-developers/zarr-python/issues/2315

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

**Comment by TomAugspurger** [comment]
One thing to figure out: what do we want for the `dtype` in the array metadata document?

```
>>> arr = zarr.create(shape=10, dtype="str")
>>> json.loads(arr.store._store_dict["zarr.json"].to_bytes())
{'shape': [10],
 'fill_value': 0,
 'chunk_grid': {'name': 'regular', 'configuration': {'chunk_shape': [10]}},
 'attributes': {},
 'zarr_format': 3,
 'data_type': 'object',
 'chunk_key_encoding': {'name': 'default',
  'configuration': {'separator': '/'}},
 'codecs': [{'name': 'bytes', 'configuration': {'endian': 'little'}}],
 'node_type': 'array',
 'storage_transformers': []}
```

Right now it's `object`. In zarr-python 2.x that was `|O`. I'm not immediately sure which is correct but I suspect either works (since presumably it's just passed to `np.dtype()` and those are the same).

File: src/zarr/core/common.py [hunk 1]
```diff
@@ -162,3 +164,10 @@ def parse_order(data: Any) -> Literal["C", "F"]:
     if data in ("C", "F"):
         return cast(Literal["C", "F"], data)
     raise ValueError(f"Expected one of ('C', 'F'), got {data} instead.")
+
+
+def parse_dtype(dtype: Any) -> np.dtype[Any]:
```
Comment (TomAugspurger) [inline src/zarr/core/common.py:169]: This type could be greatly improved. Really, we want the same signature as `np.dtype`, but with an overload for `str -> dtype(object)`. But maybe as a follow up.

File: src/zarr/core/metadata/v3.py [hunk 2]
```diff
@@ -504,6 +504,7 @@ class DataType(Enum):
     complex128 = "complex128"
     string = "string"
     bytes = "bytes"
+    object = "object"
```
[the remark this answers is not in the document]
Comment (d-v-b) [reply src/zarr/core/metadata/v3.py:507]: > Python objects are not an allowable V3 datatype.

seconding this -- we _definitely_ don't want an object dtype for v3 data.

[the remark this answers is not in the document]
Comment (TomAugspurger) [reply src/zarr/core/metadata/v3.py:507]: What behavior do people want, in terms of the in-memory representation and the on-disk metadata? Is this correct?

1. No change for `zarr_format=2`. We use object dtype in-memory and `|O` or `object` in the metadata document.
2. We use the new variable-width string handling for `zarr_format=3`. `StringDtype()` (maybe with a fallback to object if NumPy<2?) in memory and `string` in the metadata document

[the remark this answers is not in the document]
Comment (rabernat) [reply src/zarr/core/metadata/v3.py:507]: Yes that is what I want. I believe 2 is already implemented and tested for V3 here:

https://github.com/zarr-developers/zarr-python/blob/aa46b451ae6a83e1befc2525ec9629953949aa79/tests/v3/test_codecs/test_vlen.py#L49-L52

I didn't touch V2 however.

[the remark this answers is not in the document]
Comment (rabernat) [reply src/zarr/core/metadata/v3.py:507]: See also the logic in https://github.com/zarr-developers/zarr-python/blob/v3/src/zarr/core/strings.py

**Review by rabernat — APPROVED** [review]

**Comment by rabernat** [comment]
This is still not quite working with Xarray the way I was hoping. Trying to track down why.

```python
import xarray a xr
import zarr
import numpy as np

ds = xr.Dataset({"strings": ("b", np.array(["ab", "cdef", "g"], dtype=object))})
store = zarr.storage.MemoryStore({}, mode="w")
ds.to_zarr(store, zarr_version=2)
zarr.open_group(store, zarr_version=2, mode="r")["strings"].dtype
# -> dtype('<U')
```

With Zarr 2.18.3

```python
ds = xr.Dataset({"strings": ("b", np.array(["ab", "cdef", "g"], dtype=object))})
store = zarr.storage.MemoryStore({})
ds.to_zarr(store)
zarr.open_group(store, mode="r")["strings"].dtype
# -> dtype('O')
```

**Review by jhamman — APPROVED** [review]

File: tests/v3/test_v2.py [hunk 3]
```diff
@@ -44,7 +44,7 @@ def test_simple(store: StorePath) -> None:
         ("float64", 0.0),
         ("|S1", b""),
         ("|U1", ""),
-        ("object", 0),
+        ("object", ""),
```
Comment (TomAugspurger) [inline tests/v3/test_v2.py:47]: I think we agreed that this is the behavior we want. zarr-python 2.x only halfway special cased `dtype=str`. It's implicit fill value when you do `zarr.create(..., dtype=str)` was still the default for `object` (`0`).

```python
In [1]: import zarr  # 2.18.3
In [2]: arr = zarr.create(shape=10, dtype="str")

In [3]: arr.dtype
Out[3]: dtype('O')

In [4]: arr[:]
Out[4]: array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=object)
```

Comment (jhamman) [reply tests/v3/test_v2.py:47]: This looks much better!

**Comment by TomAugspurger** [comment]
@rabernat d8f24a89be7c318326ef981c04b7e3ee561444b7 fixes the issue you raised in https://github.com/zarr-developers/zarr-python/pull/2323#issuecomment-2402943872 when using this through xarray. We also need to special case `filters` when `zarr_format=2` and `dtype=str` to automatically add the vlen-utf8 codec.

With an updated `xarray-compat` branch I'll push up later + that change I get the expected behavior:

```
In [2]: import xarray as xr
   ...: import zarr
   ...: import numpy as np
   ...: 
   ...: ds = xr.Dataset({"strings": ("b", np.array(["ab", "cdef", "g"], dtype=object))})
   ...: store = zarr.storage.MemoryStore({}, mode="w")
   ...: ds.to_zarr(store, zarr_format=2)
   ...: zarr.open_group(store, zarr_format=2, mode="r")["strings"].dtype
Out[2]: dtype('O')
```

Ah, unfortunately reading the data doesn't quite work yet:

```python
In [9]: zarr.open_group(store, mode="r")["strings"][:]
```

eventually raises with

```pytb

File ~/gh/zarr-developers/zarr-python/src/zarr/codecs/_v2.py:40, in V2Compressor._decode_single(self, chunk_bytes, chunk_spec)
     38 # ensure correct dtype
     39 if str(chunk_numpy_array.dtype) != chunk_spec.dtype:
---> 40     chunk_numpy_array = chunk_numpy_array.view(chunk_spec.dtype)
     42 return get_ndbuffer_class().from_numpy_array(chunk_numpy_array)

File ~/gh/pydata/xarray/.direnv/python-3.12/lib/python3.12/site-packages/numpy/_core/_internal.py:564, in _view_is_safe(oldtype, newtype)
    561     return
    563 if newtype.hasobject or oldtype.hasobject:
--> 564     raise TypeError("Cannot change data-type for array of references.")
    565 return

TypeError: Cannot change data-type for array of references.
```

I'll look into that too.

**Comment by TomAugspurger** [comment]
Any codec experts want to chime in on whether https://github.com/zarr-developers/zarr-python/pull/2323/commits/509a5c1cb959a68043e39cc37aa50bdadcf72eb9 is appropriate? If I just remove that `.view()` we get some failures with the chunk shape not matching. So I guess we're somehow relying on that for correctness. But `.view(object)` isn't valid so I think we're OK to skip it?

**Comment by d-v-b** [comment]
> Any codec experts want to chime in on whether [509a5c1](https://github.com/zarr-developers/zarr-python/commit/509a5c1cb959a68043e39cc37aa50bdadcf72eb9) is appropriate? If I just remove that `.view()` we get some failures with the chunk shape not matching. So I guess we're somehow relying on that for correctness. But `.view(object)` isn't valid so I think we're OK to skip it?

If you leave the `.view(object)` do we get runtime errors? might be interesting to know where those are.

**Comment by TomAugspurger** [comment]
Yeah, the traceback above has the relevant bit: `TypeError: Cannot change data-type for array of references.`

**Review by jhamman — APPROVED** [review]

**Review by d-v-b — APPROVED** [review]
this looks good! As a non-string-dtype user I'm pretty surprised by the complexity involved in getting strings working in v3. Do you think we should have a section of the docs that specifically covers strings in zarr 2 / 3? That would be a separate effort from this PR of course.

**Comment by TomAugspurger** [comment]
Happy to write up those docs. One question on the intended behavior for v3. AFAICT, we don't support fixed-width, unicode strings. That uses `StringDType()`.

```python
In [1]: import zarr

In [2]: arr = zarr.create(shape=(3,), dtype="U3")

In [3]: arr[:] = ['a', 'bb', 'ccc']

In [4]: arr[:]
Out[4]: array(['a', 'bb', 'ccc'], dtype=StringDType())
```

There are some advantages to fixed-width strings when you know you have fixed-width data. Do we want to try to support that?

**Comment by rabernat** [comment]
> There are some advantages to fixed-width strings when you know you have fixed-width data. Do we want to try to support that?

Yes, I think it would be nice to support fixed-width data. In principle this can be done with just raw byte / int arrays, right? Logical vs. physical dtype etc.

**Comment by jhamman** [comment]
Can we move the fixed-width string conversation to a separate issue so we can merge this?

**Comment by TomAugspurger** [comment]
Yep, definitely.

**Pull request merged**